tutorials/shaders/compute_shaders.rst

```diff
@@ -3,70 +3,159 @@
+Invocations are instances of the shader that are running within the same
+workgroup. When we launch a compute shader from the CPU, we tell it how many
+workgroups to run. Workgroups run in parellel to each other. While running one
+workgroup, you cannot access information in another workgroup. Invocations in
+the same workgroup however, can have some limited access to other invocations.
+
+Think about workgroups and invocations as a giant nested ``for`` loop.
+
+.. code-block:: glsl
+
+    for (int x = 0; x < workgroup_size_x; x++) {
+      for (int y = 0; y < workgroup_size_y; y++) {
+         for (int z = 0; z < workgroup_size_z; z++) {
+            // Each workgroup runs independently and in parellel.
+            for (int local_x = 0; local_x < invocation_size_x; local_x++) {
+               for (int local_y = 0; local_y < invocation_size_y; local_y++) {
+                  for (int local_z = 0; local_z < invocation_size_z; local_z++) {
+                     // Compute shader runs here.
+                  }
+               }
+            }
+         }
+      }
+    }
+            
+
+Workgroups and invocations are an advanced topic, for now remember that we will
+be running two invocations per workgroup.
+
+.. code-block:: glsl
+
     // A binding to the buffer we create in our script
     layout(set = 0, binding = 0, std430) restrict buffer MyDataBuffer {
         float data[];
     }
     my_data_buffer;
 
+Here we provide information about the memory that the compute shader will have
+access to. The ``layout`` property allows us to tell the shader where to look
+for the buffer, we will need to match these ``set`` and ``binding`` positions
+from the CPU side later. 
+
+The ``restrict`` keyword tells the shader that this buffer is only going to be
+accessed from one place in this shader. In other words, we won't bind this
+buffer in another ``set`` or ``binding`` index. This is important as it allows
+the shader compiler to optimize the shader code. Always use ``restrict`` when
+you can.
+
+This is an unsized buffer meaning that it can be any size. So we need to be
+careful not to read from an index larger than the size of the buffer.
+
+.. code-block:: glsl
+
     // The code we want to execute in each invocation
     void main() {
         // gl_GlobalInvocationID.x uniquely identifies this invocation across all work groups
         my_data_buffer.data[gl_GlobalInvocationID.x] *= 2.0;
     }
 
-This code takes an array of floats, multiplies each element by 2 and store the results back in the buffer array.
- 
+Finally, we write the ``main`` function which is where all the logic happens. We
+access a position in the storage buffer using the ``gl_GlobalInvocationID``
+built in variables. ``gl_GlobalInvocationID`` gives you the global unique ID for
+the current invocation.
 
-To continue copy the code above into your newly created "compute_example.glsl" file.
+To continue copy the code above into your newly created "compute_example.glsl"
```

Comment:
I would suggest writing code instead of copy-pasting it – it's better for memorization and learning.

```suggestion
To continue, write the code above into your newly created ``compute_example.glsl``.
```